cargo-miri/bin.rs

```diff
@@ -239,9 +239,22 @@ fn xargo_check() -> Command {
 /// Execute the command. If it fails, fail this process with the same exit code.
 /// Otherwise, continue.
 fn exec(mut cmd: Command) {
-    let exit_status = cmd.status().expect("failed to run command");
-    if exit_status.success().not() {
-        std::process::exit(exit_status.code().unwrap_or(-1))
+    // On non-Unix imitate POSIX exec as closely as we can
+    #[cfg(not(unix))]
+    {
+        let exit_status = cmd.status().expect("failed to run command");
+        if exit_status.success().not() {
+            std::process::exit(exit_status.code().unwrap_or(-1))
+        }
+    }
+    // On Unix targets, actually exec
+    // If exec returns, process setup has failed. This is the same error condition as the expect in
+    // the non-Unix case.
+    #[cfg(unix)]
+    {
+        use std::os::unix::process::CommandExt;
+        let error = cmd.exec();
```

Comment:
Oh, that's a neat extension.

But note the doc comment of this function: it promises to return to the caller when the program is done. *Some* callers rely on this.

Reply:
At least this invocation relies on stuff happening after `exec`:

https://github.com/rust-lang/miri/blob/302e9ae206ed148fc009d84c05ce0c99877ef6aa/cargo-miri/bin.rs#L958

Reply:
Ah, there is also this:

https://github.com/rust-lang/miri/blob/302e9ae206ed148fc009d84c05ce0c99877ef6aa/cargo-miri/bin.rs#L873-L876

and just dumping the JSON earlier leads to errors here. So it's not that easy to reorder them.

Reply:
Looks like rustdoc passes something like `"-o" "/tmp/rustdoctest4SiV4X/rust_out"`. Which will be overwritten by rustc even when we just do "--emit=metadata".
So to reorder the JSON dumping to before the exec, we have to also adjust the `-o` such that rustc won't overwrite the JSON.

Reply:
I think a solution here is to only swap the `phase_runner` calls to this sort of implementation. At least the tests pass if I do that for this specific call: https://github.com/rust-lang/miri/blob/302e9ae206ed148fc009d84c05ce0c99877ef6aa/cargo-miri/bin.rs#L1075

I think the most correct thing would be to swap the `RustDoc` branch right next to that to proper exec as well, but I think to do that I need to write to the current process's stdin, and I'm trying to avoid `unsafe { libc::write(libc::STDIN_FILENO) }` but it seems tricky, if even possible. The standard library really doesn't seem to be designed for writing to stdin.

Reply:
> I think the most correct thing would be to swap the RustDoc branch right next to that to proper exec as well, but I think to do that I need to write to the current process's stdin

Wait, why does the `stdin` thing that `exec_with_pipe` does not work?

Reply:
There is no child process to get a writable `stdin` handle for, if we're going to do `cmd.exec()`.

Reply:
Oh I see. Is there a way to create a pipe in advance, fill it with the input, and use `Stdio: From<OwnedFd>`? The only other idea I have is to create a temporary file...

Reply:
> I think a solution here is to only swap the phase_runner calls to this sort of implementation.

I think it'd be nicer to consistently do proper `exec` everywhere, but yes that would also work.

Reply:
Your comments above are helping me understand how a lot of cargo-miri works, I'm going to see if I can do proper `exec` everywhere.

Reply:
If you can think of how to extend the existing comments to better explain how cargo-miri works, that'd be great. :)

(Also beware of conflicts with some in-flight PRs, probably best to rebase in around 50min before you do a lot of work.)